ttnn/cpp/ttnn/operations/conv/conv_transpose2d/conv_transpose2d.cpp

```diff
@@ -0,0 +1,350 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "conv_transpose2d.hpp"
+#include <sys/types.h>
+#include <cstdint>
+#include "common/bfloat16.hpp"
+
+using namespace tt;
+namespace ttnn {
+namespace operations::conv {
+using sliding_window::SlidingWindowConfig;
+using sliding_window::ParallelConfig;
+
+namespace conv_transpose2d {
+
+template <typename T>
+Tensor _transform_weights_for_conv_transpose2d(
+    const Tensor& conv_weight_tensor) {
+    auto in_w_shape = conv_weight_tensor.get_legacy_shape();
+    auto dtype = conv_weight_tensor.dtype();
+    // in_w_shape = {in_channels, out_channels, kernel_height, kernel_width}
+    // out_w_shape = {out_channels, in_channels, kernel_height, kernel_width}
+    //Flip kernel_height and kernel_width
+    auto compute = [&in_w_shape, &dtype](const auto& input_buffer) {
+        auto in_channels   = in_w_shape[0];
+        auto out_channels  = in_w_shape[1];
+        auto kernel_height = in_w_shape[2];
+        auto kernel_width  = in_w_shape[3];
+        ttnn::SimpleShape output_shape{out_channels, in_channels, kernel_height, kernel_width};
+        auto output_buffer = owned_buffer::create<T>(output_shape.volume());
+
+        for(auto in_kernel_height_index = 0; in_kernel_height_index < kernel_height; in_kernel_height_index++)  {
+            auto out_buffer_kh_index = kernel_height - in_kernel_height_index - 1;
+            for(auto in_kernel_width_index = 0; in_kernel_width_index < kernel_width; in_kernel_width_index++)  {
+                auto out_buffer_kw_index = kernel_width - in_kernel_width_index - 1;
+                for(auto in_channels_index = 0; in_channels_index < in_channels; in_channels_index++)   {
+                    for(auto out_channels_index = 0; out_channels_index < out_channels; out_channels_index++)   {
+                        auto in_idx = in_channels_index * kernel_height * kernel_width * out_channels + out_channels_index * kernel_height * kernel_width + in_kernel_height_index * kernel_width + in_kernel_width_index;
+                        auto out_idx = out_channels_index * in_channels * kernel_height * kernel_width + in_channels_index * kernel_height * kernel_width + out_buffer_kh_index * kernel_width + out_buffer_kw_index;
+                        output_buffer[out_idx] = input_buffer[in_idx];
+                    }
+                }
+            }
+        }
```

Comment:
Maybe smth like this. It avoids redundant multiplication and will probably improve data locality.

```
for (auto out_channels_index = 0; out_channels_index < out_channels; out_channels_index++) {
    auto out_base_idx = out_channels_index * in_channels * kernel_height * kernel_width;
    for (auto in_channels_index = 0; in_channels_index < in_channels; in_channels_index++) {
        auto in_base_idx = in_channels_index * kernel_height * kernel_width * out_channels + out_channels_index * kernel_height * kernel_width;
        for (auto in_kernel_height_index = 0; in_kernel_height_index < kernel_height; in_kernel_height_index++) {
            auto out_buffer_kh_index = kernel_height - in_kernel_height_index - 1;
            auto in_height_offset = in_kernel_height_index * kernel_width;
            auto out_height_offset = out_buffer_kh_index * kernel_width;
            for (auto in_kernel_width_index = 0; in_kernel_width_index < kernel_width; in_kernel_width_index++) {
                auto out_buffer_kw_index = kernel_width - in_kernel_width_index - 1;
                
                auto in_idx = in_base_idx + in_height_offset + in_kernel_width_index;
                auto out_idx = out_base_idx + out_height_offset + out_buffer_kw_index;

                output_buffer[out_idx] = input_buffer[in_idx];
            }
        }
    }
}
```

Reply:
Done.